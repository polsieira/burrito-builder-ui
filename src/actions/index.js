export const setOrders = orders => ({
  type: 'SET_ORDERS',
  orders
});

export const addOrders = newOrder => ({
  type: 'ADD_ORDERS',
  newOrder
});
