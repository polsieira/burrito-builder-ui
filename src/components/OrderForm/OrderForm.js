import React, { Component } from 'react';
import { addOrders } from '../../actions';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { addNewOrder } from '../../apiCalls';

export class OrderForm extends Component {
  constructor() {
    super();
    this.state = {
      name: '',
      ingredients: [],
      price: 0
    };
  }

  handleNameChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  }

  handleIngredientChange = e => {
    e.preventDefault();
    if (!this.state.ingredients.includes(e.target.name)) {
      this.setState({ ingredients: [...this.state.ingredients, e.target.name] });
      console.log(e.target.dataset.price)
      this.setState({ price: this.state.price + parseFloat(e.target.dataset.price) });
    }
  }

  handleSubmit = async e => {
    e.preventDefault();
    if (this.state.name && this.state.ingredients.length > 0) {
      if (this.state.name === 'Robbie' || 'Travis') {
        throw new Error('Sorry your burrito sucks')
      }
      try {
        const data = await addNewOrder(this.state);
        this.props.addOrders(data);
      } catch ({ message }) {
        // do something with error action
      }
      this.clearInputs();
    }
  }

  clearInputs = () => {
    this.setState({ name: '', ingredients: [] });
  }

  render() {
    const possibleIngredients = [{ name: 'beans', price: .99 }, { name: 'steak', price: .99 }, { name: 'carnitas', price: .99 }, { name: 'sofritas', price: .99 }, { name: 'lettuce', price: .10 }, { name: 'queso fresco', price: .99 }, { name: 'pico de gallo', price: .99 }, { name: 'hot sauce', price: .10 }, { name: 'guacamole', price: 1.99 }, { name: 'jalapenos', price: .99 }, { name: 'cilantro', price: .10 }, { name: 'sourcream', price: .99 }];
    const ingredientButtons = possibleIngredients.map(ingredient => {
      return (
        <button key={ingredient.name} name={ingredient.name} data-price={ingredient.price} onClick={e => this.handleIngredientChange(e)}>
          {ingredient.name}
        </button>
      )
    });

    return (
      <form>
        <input
          type='text'
          placeholder='Name'
          name='name'
          value={this.state.name}
          onChange={e => this.handleNameChange(e)}
        />

        {ingredientButtons}

        <p>Order: {this.state.ingredients.join(', ') || 'Nothing selected'}</p>

        <button className='test-this-button' onClick={e => this.handleSubmit(e)}>
          Submit Order
        </button>
      </form>
    )
  }
}

export const mapDispatchToProps = dispatch => {
  return bindActionCreators({
    addOrders
  }, dispatch);
};

export default connect(null, mapDispatchToProps)(OrderForm);