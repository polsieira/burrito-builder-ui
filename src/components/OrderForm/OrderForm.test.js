import React from 'react';
import { shallow } from 'enzyme';
import { OrderForm, mapDispatchToProps } from './OrderForm';
import { addOrders } from '../../actions';

jest.mock('../../apiCalls.js');

// Sorry about the nested describe blocks, I like them

describe('Orders', () => {

  describe('Orders container', () => {
    let wrapper;
    const mockAddOrders = jest.fn();

    beforeEach(() => {
      wrapper = shallow(<OrderForm addOrders={mockAddOrders} />);
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should update name in state when handleNameChange is called', () => {
      const mockEvent = {
        target: {
          name: 'name',
          value: 'Pol'
        },
        preventDefault: jest.fn()
      };

      const expected = 'Pol';

      wrapper.instance().handleNameChange(mockEvent);


      expect(wrapper.state('name')).toEqual(expected);
    });

    it('should update ingredients in state when handleIngredientChange is called', () => {
      const mockEvent = {
        target: {
          name: 'BEANS',
        },
        preventDefault: jest.fn()
      };
      wrapper.setState({
        ingredients: [
          'beans',
          'BeAnS'
        ]
      })
      const expected = [
        'beans',
        'BeAnS',
        'BEANS'
      ];

      wrapper.instance().handleIngredientChange(mockEvent);


      expect(wrapper.state('ingredients')).toEqual(expected);
    });

    it('should clear all variables in state when clearInputs is invoked', () => {
      const expected = {
        name: '',
        ingredients: [],
      };

      wrapper.instance().setState({
        name: 'Pol',
        ingredients: [
          'beans',
          'BeAnS',
          'BEANS'
        ]
      });

      wrapper.instance().clearInputs();

      expect(wrapper.state()).toEqual(expected);
    });

    it('should call handleSubmit on click', () => {
      wrapper.instance().handleSubmit = jest.fn();
      wrapper.instance().forceUpdate();

      wrapper.find('.test-this-button').simulate('click');

      expect(wrapper.instance().handleSubmit).toHaveBeenCalled();
    });

    it.skip('should try a fetch and call addOrders on invocation of handleSubmit', () => {

    });

  });

  describe('mapDispatchToProps', () => {

    const mockDispatch = jest.fn();



  });

});