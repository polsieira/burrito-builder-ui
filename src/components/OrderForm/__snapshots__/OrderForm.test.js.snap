// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orders Orders container should match the snapshot 1`] = `
<form>
  <input
    name="name"
    onChange={[Function]}
    placeholder="Name"
    type="text"
    value=""
  />
  <button
    data-price={0.99}
    key="beans"
    name="beans"
    onClick={[Function]}
  >
    beans
  </button>
  <button
    data-price={0.99}
    key="steak"
    name="steak"
    onClick={[Function]}
  >
    steak
  </button>
  <button
    data-price={0.99}
    key="carnitas"
    name="carnitas"
    onClick={[Function]}
  >
    carnitas
  </button>
  <button
    data-price={0.99}
    key="sofritas"
    name="sofritas"
    onClick={[Function]}
  >
    sofritas
  </button>
  <button
    data-price={0.1}
    key="lettuce"
    name="lettuce"
    onClick={[Function]}
  >
    lettuce
  </button>
  <button
    data-price={0.99}
    key="queso fresco"
    name="queso fresco"
    onClick={[Function]}
  >
    queso fresco
  </button>
  <button
    data-price={0.99}
    key="pico de gallo"
    name="pico de gallo"
    onClick={[Function]}
  >
    pico de gallo
  </button>
  <button
    data-price={0.1}
    key="hot sauce"
    name="hot sauce"
    onClick={[Function]}
  >
    hot sauce
  </button>
  <button
    data-price={1.99}
    key="guacamole"
    name="guacamole"
    onClick={[Function]}
  >
    guacamole
  </button>
  <button
    data-price={0.99}
    key="jalapenos"
    name="jalapenos"
    onClick={[Function]}
  >
    jalapenos
  </button>
  <button
    data-price={0.1}
    key="cilantro"
    name="cilantro"
    onClick={[Function]}
  >
    cilantro
  </button>
  <button
    data-price={0.99}
    key="sourcream"
    name="sourcream"
    onClick={[Function]}
  >
    sourcream
  </button>
  <p>
    Order: 
    Nothing selected
  </p>
  <button
    className="test-this-button"
    onClick={[Function]}
  >
    Submit Order
  </button>
</form>
`;
